### PR TITLE
Receive a message and post a response to the same channel

### PR DESCRIPTION
**Pull request by kyle-yee-74**

### CODE REVIEW

File: discord_bot/discord_bot.py [hunk 1]
```diff
@@ -144,24 +142,45 @@ def load_config(config_file_path):
                     self.logger.warning("Invalid Session")
                 elif message["op"] == Opcodes.DISPATCH:
                     event = message["t"]
-                    if event == Events.READY:
+                    if event == Events.READY.value:
```
Comment (xelaadryth) [inline discord_bot/discord_bot.py]: Doesn't this already compare by value? The purpose of an enum is to make it as easy as possible to compare values intuitively without adding additional syntax, so we should probably remove the `.value` from all if cases.

File: discord_bot/discord_bot.py [hunk 2]
```diff
@@ -144,24 +142,45 @@ def load_config(config_file_path):
...
                         self.session_id = message["d"]["session_id"]
+                    elif event == Events.MESSAGE_CREATE.value:
+                        channel_id = message["d"]["channel_id"]
```
Comment (xelaadryth) [inline discord_bot/discord_bot.py]: I think I'd rather the separate function take the raw message completely and parse it itself to save as many lines as possible in this mega-function.

File: discord_bot/discord_bot.py [hunk 3]
```diff
@@ -144,24 +142,45 @@ def load_config(config_file_path):
...
-    #         }).json()
+    async def respond_message(self, channel_id, author, content):
+        """
+        Receives a message and responds to the user in the same channel.
+        :param channel_id: int - channel id
+        :param author: string - author of the message
+        :param content: string - message the user inputs
+        :return:
+        """
+        defaults = {
```
Comment (xelaadryth) [inline discord_bot/discord_bot.py]: This is incredibly confusing and also extremely weird. You're defining a dictionary, then redefining another dictionary based on it partially with keyword arguments and partially copying another dictionary, and then passing it as a different keyword argument just to make it more confusing.

I heavily advise you to never use `kwargs` unless you know exactly what you're doing and why you need to do it; you should be explicit with your arguments.

Here's an example of how it should look:
http://aiohttp.readthedocs.io/en/stable/client.html#custom-headers

File: discord_bot/discord_bot.py [hunk 4]
```diff
@@ -144,24 +142,45 @@ def load_config(config_file_path):
...
+                "User-Agent": "MomoiBot (https://github.com/ShifuYee/Discord-Bot)"
+            }
+        }
+        kwargs = dict(defaults, data={
+            "content": f"{author} how may I help?"
+        })
+        if content.startswith("!help"):
+            url = "{}/channels/{}/messages".format(self.config["discord_api_endpoint"], channel_id)
+            async with ClientSession() as session:
+                async with session.request(
```
Comment (xelaadryth) [inline discord_bot/discord_bot.py]: If it easily fits on one line, you should probably leave it as one line, or it looks more complicated than it needs to be.

File: discord_bot/events.py [hunk 5]
```diff
@@ -1,7 +1,12 @@
 from enum import Enum
 
 
-class Events (Enum):
+class NoValue(Enum):
```
Comment (xelaadryth) [inline discord_bot/events.py]: What's the purpose of removing the assigned value from the enumeration? This complicates the code without adding additional functionality, and even makes it harder to debug issues if you need to know the assigned value. I recommend removing this NoValue class since it provides us with...no value ;3

File: discord_bot/discord_bot.py [hunk 6]
```diff
@@ -144,24 +142,45 @@ def load_config(config_file_path):
...
+        Receives a message and responds to the user in the same channel.
+        :param channel_id: int - channel id
+        :param author: string - author of the message
+        :param content: string - message the user inputs
+        :return:
+        """
+        defaults = {
+            "headers": {
+                "Authorization": "Bot {}".format(self.config["handshake_identity"]["token"]),
+                "User-Agent": "MomoiBot (https://github.com/ShifuYee/Discord-Bot)"
```
Comment (xelaadryth) [inline discord_bot/discord_bot.py]: We should add the user-agent field to the config file as well, especially since it will be repeated in multiple places.

**Review by xelaadryth — CHANGES_REQUESTED** [review]

File: discord_bot/discord_bot.py (hunk 1 above)
Comment (kyle-yee-74) [reply discord_bot/discord_bot.py]: Going through the debugger. 
I found that Events.READY actually doesn't equal 'READY'.
https://docs.python.org/3/library/enum.html
8.13.14.1.3. Using a descriptive string
This example is why I had to use `.value` for all cases.

Comment (xelaadryth) [reply discord_bot/discord_bot.py]: See https://docs.python.org/3/library/enum.html#others

Time to use a little mix-in magic to make it work, but basically your issue is that the way it compares equals (I believe the hidden method is called `__eq__`) is not using the `str` method. If you use a `str` mix-in then it should work properly.

File: discord_bot/discord_bot.py [hunk 7]
```diff
@@ -144,24 +142,39 @@ def load_config(config_file_path):
                     self.logger.warning("Invalid Session")
                 elif message["op"] == Opcodes.DISPATCH:
                     event = message["t"]
```
Comment (xelaadryth) [inline discord_bot/discord_bot.py]: I actually might move all of this if-casing into a separate method called `dispatch_handler` or `event_handler` or something, whatever makes the most sense, to get `gateway_handler` to be smaller.

File: discord_bot/discord_bot.py [hunk 8]
```diff
@@ -144,24 +142,39 @@ def load_config(config_file_path):
...
-    #     return requests.post(
-    #         "{}/channels/{}/messages".format(
-    #             self.config["discord_api_endpoint"], channel["id"]),
-    #         json={
-    #             "content": content
-    #         }).json()
+    async def respond_message(self, message):
+        """
+        Receives a message, parses the message, and responds to the user in the same channel.
+        :param message: obj - the payload of the MESSAGE_CREATE event
```
Comment (xelaadryth) [inline discord_bot/discord_bot.py]: Everything is an `obj`, be more specific about the type. I believe it's a `dict` but also talk about what keys it should/can expect.

File: discord_bot/discord_bot.py [hunk 9]
```diff
@@ -144,24 +142,39 @@ def load_config(config_file_path):
...
+        :return: string - defined by data
```
Comment (xelaadryth) [inline discord_bot/discord_bot.py]: The class name is `str` not `string`, but also this method doesn't actually return a string. `r.json()` return a "Future", which is something you can wait on to return an actual value. `await X` returns the actual value, which I believe is a `dict`?

File: discord_bot/discord_bot.py [hunk 10]
```diff
@@ -144,24 +142,39 @@ def load_config(config_file_path):
...
+                            "{}/channels/{}/messages".format(self.config["discord_api_endpoint"], channel_id),
+                            headers={
+                                "Authorization": "Bot {}".format(self.config["handshake_identity"]["token"]),
+                                "User-Agent": "{}".format(self.config["user_agent"])
+                            },
+                            data={
+                                "content": f"<@{author_id}> how may I help?"
+                            }) as r:
+                        if r.status == HTTPStatus.OK:
+                            return await r.json()
```
Comment (xelaadryth) [inline discord_bot/discord_bot.py]: Right now we don't do anything with this json response, you could probably add a `TODO` on it to do something with it.

File: discord_bot/discord_bot.py [hunk 11]
```diff
@@ -144,24 +142,39 @@ def load_config(config_file_path):
...
+                        else:
+                            raise DiscordBotException(
+                                f"Expected a post to discord channel {channel_id},"
```
Comment (xelaadryth) [inline discord_bot/discord_bot.py]: We're making the post, so you might mean "Expected a successful response when POSTing to discord channel..."

File: discord_bot/discord_bot.py [hunk 12]
```diff
@@ -144,24 +142,39 @@ def load_config(config_file_path):
...
+        content = message["d"]["content"]
+        if author_id != self.config["momoi_id"]:
+            if content.startswith("!help"):
+                async with ClientSession() as session:
+                    async with session.request(
+                            "POST",
+                            "{}/channels/{}/messages".format(self.config["discord_api_endpoint"], channel_id),
+                            headers={
+                                "Authorization": "Bot {}".format(self.config["handshake_identity"]["token"]),
+                                "User-Agent": "{}".format(self.config["user_agent"])
```
Comment (xelaadryth) [inline discord_bot/discord_bot.py]: You don't need to format anything in, you can just use `self.config["user_agent"]` directly.

File: discord_bot/discord_bot.py [hunk 13]
```diff
@@ -144,24 +142,39 @@ def load_config(config_file_path):
...
+        """
+        channel_id = message["d"]["channel_id"]
+        author_id = message["d"]["author"]["id"]
+        content = message["d"]["content"]
+        if author_id != self.config["momoi_id"]:
+            if content.startswith("!help"):
+                async with ClientSession() as session:
+                    async with session.request(
+                            "POST",
+                            "{}/channels/{}/messages".format(self.config["discord_api_endpoint"], channel_id),
```
Comment (xelaadryth) [inline discord_bot/discord_bot.py]: You can do this in a separate PR, but we should probably convert all of our `"blah{}".format(myVar)` to `f"blah{myVar}"` for consistency

File: discord_bot/discord_bot.py [hunk 14]
```diff
@@ -144,24 +142,39 @@ def load_config(config_file_path):
...
+        :param message: obj - the payload of the MESSAGE_CREATE event
+        :return: string - defined by data
+        """
+        channel_id = message["d"]["channel_id"]
+        author_id = message["d"]["author"]["id"]
+        content = message["d"]["content"]
+        if author_id != self.config["momoi_id"]:
+            if content.startswith("!help"):
+                async with ClientSession() as session:
+                    async with session.request(
```
Comment (xelaadryth) [inline discord_bot/discord_bot.py]: This line of code is crazy long and hard to read, let's store everything in variables like `url`, `headers`, and `payload` on separate lines so that this line can be a simple `async with session.request("POST", url, headers=headers, data=payload) as r:` in one line.

If we cared about speed and memory efficiency then we could do it in one line, but if we cared about speed we'd be using C++ not Python anyway :p

**Review by xelaadryth — CHANGES_REQUESTED** [review]

File: discord_bot/discord_bot.py [hunk 15]
```diff
@@ -146,22 +153,37 @@ def load_config(config_file_path):
                     event = message["t"]
```
Comment (xelaadryth) [inline discord_bot/discord_bot.py]: My other comment got hidden, but we should probably move all of this dispatch/event handling into its own separate method to let `gateway_handler` be smaller.

File: discord_bot/discord_bot.py [hunk 16]
```diff
@@ -121,6 +128,17 @@ def load_config(config_file_path):
...
+    async def dispatch_handler(self, message):
+        """
+        Handles all types of DISPATCH events.
+        :param message: dict - the message object.
+        """
+        event = message["t"]
+        if event == Events.READY:
+            self.session_id = message["d"]["session_id"]
+        elif event == Events.MESSAGE_CREATE:
+            await self.respond_message(message)
```
Comment (xelaadryth) [inline discord_bot/discord_bot.py:140]: Actually, shouldn't we `asyncio.ensure_future`? `await` will block the bot from doing anything until it finishes processing.

**Review by xelaadryth — APPROVED** [review]
If you fix that one minor thing and test it, feel free to merge.
